utils.go

```diff
@@ -12,80 +11,44 @@ import (
-	// channels.
-	if !IsPublicNode(channels, channel.PubKeyBytes) {
-		log.Debugf("Skipping channel %v due to "+
-			"counterparty %x being unadvertised",
-			channel.ChannelID, channel.PubKeyBytes[:])
-		return nil, false
-	}
-
-	// Now, we'll need to determine which is the correct policy for HTLCs
-	// being sent from the remote node.
-	var remotePolicy *lndclient.RoutingPolicy
-	if bytes.Equal(channel.PubKeyBytes[:], chanInfo.Node1[:]) {
-		remotePolicy = chanInfo.Node1Policy
-	} else {
-		remotePolicy = chanInfo.Node2Policy
-	}
-
-	return remotePolicy, true
-}
-
-func SelectHopHints(ctx context.Context, lnd *lndclient.LndServices, amtMSat btcutil.Amount,
-	numMaxHophints int, includeNodes map[route.Vertex]struct{}) ([][]zpay32.HopHint, error) {
+// SelectHopHints is a direct port of the SelectHopHints found in lnd. It
+// iterates through all the active and public channels available and returns
+// eligible channels. Eligibility requirements are simple: does the channel
+// have enough liquidity to fulfill the request and is the node whitelisted (if // specified)
+func SelectHopHints(ctx context.Context, lnd *lndclient.LndServices,
+	amtMSat btcutil.Amount, numMaxHophints int,
+	includeNodes map[route.Vertex]struct{}) ([][]zpay32.HopHint, error) {
 
-	openChannels, err := lnd.Client.ListChannels(ctx)
+	// Fetch all active and public channels.
+	openChannels, err := lnd.Client.ListChannels(ctx, true, true)
 	if err != nil {
 		return nil, err
 	}
+
 	// We'll add our hop hints in two passes, first we'll add all channels
 	// that are eligible to be hop hints, and also have a local balance
 	// above the payment amount.
 	var totalHintBandwidth btcutil.Amount
 	hopHintChans := make(map[int]*lndclient.ChannelEdge)
 	hopHints := make([][]zpay32.HopHint, 0, numMaxHophints)
+
 	for i, channel := range openChannels {
 		// In this first pass, we'll ignore all channels in
 		// isolation can't satisfy this payment.
-		if channel.RemoteBalance < amtMSat {
-			continue
-		}
-
-		// If includeNodes is set, we'll only add channels with peers in include_node.
-		// This is done to respect the last_hop parameter
-		if _, ok := includeNodes[channel.PubKeyBytes]; includeNodes != nil && !ok {
-			continue
-		}
 
 		// Retrieve extra info for each channel not available in listChannels
 		chanInfo, err := lnd.Client.GetChanInfo(ctx, channel.ChannelID)
 		if err != nil {
 			return nil, err
 		}
-
-		// If this channel can't be a hop hint, then skip it.
-		edgePolicy, canBeHopHint := chanCanBeHopHint(openChannels, channel, chanInfo)
-		if edgePolicy == nil || !canBeHopHint {
+		// Cache the GetChanInfo result since it might be useful
+		hopHintChans[i] = chanInfo
+		if channel.RemoteBalance < amtMSat {
+			continue
+		}
+		// If includeNodes is set, we'll only add channels with peers in
+		// include_node. This is done to respect the last_hop parameter
+		if _, ok := includeNodes[channel.PubKeyBytes]; includeNodes == nil && len(includeNodes) != 0 && !ok {
```

Comment:
I think that we can squash this with the original commit? Unrelated to performance enhancements.